### PR TITLE
Display hints on the right elements 

### LINKED ISSUE #487: Tridactyl doesn't display hints on the right elements on some websites
Sometimes anchor elements contain other elements that are displayed far from the anchor. This happens on https://jupyter.org/try . If the first child of an anchor element isn't a textNode, Tridactyl should probably place hints on the child rather than on the anchor element.

### PR DESCRIPTION
**Pull request by MatiasStorm**
This fixes #487.

### CODE REVIEW

**Comment by bovine3dom** [comment]
Thanks for this. I'll check that this works it this afternoon and merge if it does.

A little unsolicited advice first: you can avoid commits like 074cee7bb6f1d54f78e2088a3136d35551ad5b86 which don't really add any information by doing something like:

```
git remote add upstream git clone git@github.com:tridactyl/tridactyl.git # do this once
git pull --rebase upstream/master # do this when you want to update your repository
git push --force
```

In general it's a little bit anti-social to do a force push, but it's OK to be anti-social if you're working on your own :)

File: src/content/hinting.ts [hunk 1]
```diff
@@ -844,6 +844,14 @@ export function hintables(selectors = DOM.HINTTAGS_selectors, withjs = false) {
         const elemSet = new Set([...elems, ...DOM.hintworthy_js_elems])
         elems = [...elemSet]
     }
+    elems.forEach((element, index) => {
```
Comment (bovine3dom) [inline src/content/hinting.ts:847]: Could you add a note (either in a comment or a commit message) which issue this fixes?

Just from looking at the code alone it isn't obvious why it is needed (which is probably why the code wasn't here initially and we had the issue :) )

**Comment by bovine3dom** [comment]
More unsolicited `git` advice: usually people write commit messages in the "imperative voice", i.e. as if the commit message is an order to the computer. There is some logic to this in that each commit is a patch which is set of instructions for the computer: e.g.

> Attach hints to child divs or spans

is what 49ff5ed tells the computer to do, so that is what its commit message should be.

I care only mildly about this for consistency. Some people get a bit snobby about it, so I thought I should point it out :)

**Comment by bovine3dom** [comment]
There's a small regression here on www.bbc.co.uk/news:



The top image is without this PR, the bottom is with this PR. You can see that the hint for "Pressure grows on PM top aide over lockdown trip" is now missing / moved.

For this exact example, it doesn't matter - HO points to the same place.

If you scroll down the page, to the "Must See" section, the video links there have the same problem except that the remaining links don't work (so the ones that are now missing / hidden were important).

I'm reluctant to merge this unless we can fix this issue - I'm not aware of many sites that suffer from #487 and the BBC is one of the most visited sites in the world (it is also one of the most visited sites in my flat... ;) ), so I don't want to break it.

**Comment by MatiasStorm** [comment]
Thank you very much for the quick reply and the amazing feedback!
As you can probably tell I'm pretty new to this. So I really appreciate you taking your time to share your advice. 

I have fix all the issues you pointed out and tested it on www.bbc.co.uk/news, and it looks fine to me :)

**Comment by MatiasStorm** [comment]
Quick question about linting:
I have tried running both ```tslint --project .``` and ```tslint --config ./tslint.json --project .``` but none of them give the same linting errors as ```ci/circleci```.

Do you know what I could be doing wrong? My ```tslint``` version is ```6.1.2```

**Comment by bovine3dom** [comment]
You should be using tslint 5.x: https://github.com/tridactyl/tridactyl/blob/ced64e8fb689a0a4efb5cdccfa0ff642965e602b/package.json#L41

You can do this easily with `$(yarn bin)/tslint` - it'll use the package defined in package.json.

**Comment by bovine3dom** [comment]
Thanks for fixing this. I'll check it again tomorrow and, with luck, merge it.

If you are feeling brave you could "squash" the commits to make the history more useful - we don't really care that you had linting problems, so the commits that refer to those could be squashed into the previous ones. Ideally, each commit should be "interesting". That makes tools like `git blame` and `git bisect` more useful when we use them in the future.

The git book has some information on the mechanics of squashing here: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_changing_multiple 

I would probably personally squash it such that we only had commits relating to 0a4b340bfabc902c4cb1346cc18b0be1b3ba1c75 and 49ff5ede2b78aff86f2d08748dac49b1372fd3f6 - i.e. your initial attempt and then your fix for the BBC site.

**Comment by MatiasStorm** [comment]
I have managed to squash the commits, and hope that everything looks and works as it is supposed to.
Thanks again for all of your help, I think I'm gonna contribute a lot more to this project in the future! (Hopefully with fewer mistakes ;-) )

**Comment by bovine3dom** [comment]
> I think I'm gonna contribute a lot more to this project now!

That'd be fantastic!

If you're looking for a fairly big project with some mentoring, we made https://tridactyl.xyz/ideas/ for GSoC. We weren't accepted on to GSoC but we'd still be happy to do a GSoC-style thing with you.

Otherwise there are plenty of issues open on the repository. Some of them are really tough :(

**Pull request merged**